lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventTrackerImpl.java

```diff
@@ -0,0 +1,187 @@
+        updateInternal(true, update);
+    }
+
+    /**
+     * Success case, where we add the events from the update to the current state.
+     */
+    private void accumulateStateWith(LockWatchStateUpdate update) {
+        updateInternal(false, update);
+    }
+
+    private void updateInternal(boolean resetState, LockWatchStateUpdate update) {
+        TreeRangeSet<LockDescriptor> watchesToUpdate;
+        Map<LockDescriptor, LockWatchInfo> locksToUpdate;
+        int firstUpdateIndex;
+
+        if (resetState) {
+            openLocksEvents.clear();
+            watchesToUpdate = TreeRangeSet.create();
+            locksToUpdate = new HashMap<>();
+            firstUpdateIndex = 0;
+        } else {
+            watchesToUpdate = TreeRangeSet.create(watches.get());
+            locksToUpdate = new HashMap<>(singleLocks.get());
+            firstUpdateIndex = Ints.saturatedCast(lastKnownVersion.getAsLong() + 1 - update.events().get(0).sequence());
+        }
+
+        LockWatchStateUpdater visitor = new LockWatchStateUpdater(watchesToUpdate, locksToUpdate, openLocksEvents);
+        update.events().listIterator(firstUpdateIndex).forEachRemaining(event -> event.accept(visitor));
+        setAll(watchesToUpdate, locksToUpdate, update);
+    }
+
+    private void setAll(TreeRangeSet<LockDescriptor> updatedWatches, Map<LockDescriptor, LockWatchInfo> updatedLocks,
+            LockWatchStateUpdate update) {
+        lock.writeLock().lock();
+        try {
+            watches.set(updatedWatches);
+            singleLocks.set(updatedLocks);
+            lastKnownVersion = update.lastKnownVersion();
+            leaderId = update.leaderId();
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private boolean updateNotPossible(LockWatchStateUpdate update, long currentTrackedVersion) {
+        return update.events().isEmpty()
+                || staleUpdate(update, currentTrackedVersion)
+                || versionDiscontinuity(update, currentTrackedVersion);
+    }
+
+    private boolean staleUpdate(LockWatchStateUpdate update, long currentTrackedVersion) {
+        return update.lastKnownVersion().getAsLong() <= currentTrackedVersion;
+    }
+
+    private boolean versionDiscontinuity(LockWatchStateUpdate update, long currentTrackedVersion) {
+        return update.events().get(0).sequence() > currentTrackedVersion + 1;
+    }
+
+    @Override
+    public void setLockWatchStateForStartTimestamp(long startTimestamp, VersionedLockWatchState lockWatchState) {
+        startTsToLockWatchState.put(startTimestamp, lockWatchState);
+    }
+
+    @Override
+    public VersionedLockWatchState getLockWatchStateForStartTimestamp(long startTimestamp) {
+        return startTsToLockWatchState.get(startTimestamp);
+    }
+
+    @Override
+    public VersionedLockWatchState removeLockWatchStateForStartTimestamp(long startTimestamp) {
+        return startTsToLockWatchState.remove(startTimestamp);
```

Comment:
How is this going to be used?

Reply:
You could, but then you force your transactions to abort. Like this you keep it in memory until you get to commit time, when it gets removed

Reply:
Oh wait, I just noticed that this map is for startTimestamp -> state